Support for OpenBSD / NetBSD

Hi,

I am trying to install agent on both OpenBSD / NetBSD (cirruslabs/cirrus-ci-docs#970). Right now agent's tests are failing with:

On **OpenBSD**:
```
?   	github.com/cirruslabs/cirrus-ci-agent/api	[no test files]
# github.com/mitchellh/go-ps
../go/pkg/mod/github.com/mitchellh/go-ps@v1.0.0/process.go:31:9: undefined: processes
../go/pkg/mod/github.com/mitchellh/go-ps@v1.0.0/process.go:39:9: undefined: findProcess
ok  	github.com/cirruslabs/cirrus-ci-agent/cmd/agent	(cached)
?   	github.com/cirruslabs/cirrus-ci-agent/cmd/metrics	[no test files]
ok  	github.com/cirruslabs/cirrus-ci-agent/internal/cirrusenv	(cached)
?   	github.com/cirruslabs/cirrus-ci-agent/internal/client	[no test files]
FAIL	github.com/cirruslabs/cirrus-ci-agent/internal/executor [build failed]
--- FAIL: TestMetrics (0.00s)
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x875e04]

goroutine 6 [running]:
testing.tRunner.func1.2({0x8da540, 0xd30290})
	/usr/local/go/src/testing/testing.go:1209 +0x339
testing.tRunner.func1(0xc000091a00)
	/usr/local/go/src/testing/testing.go:1212 +0x256
panic({0x8da540, 0xd30290})
	/usr/local/go/src/runtime/panic.go:1038 +0x25b
github.com/cirruslabs/cirrus-ci-agent/internal/executor/metrics_test.TestMetrics(0xc000091a00)
	/home/postgres/cirrus-ci-agent/internal/executor/metrics/metrics_test.go:20 +0xe4
testing.tRunner(0xc000091a00, 0x988b18)
	/usr/local/go/src/testing/testing.go:1259 +0x102
created by testing.(*T).Run
	/usr/local/go/src/testing/testing.go:1306 +0x3ab
FAIL	github.com/cirruslabs/cirrus-ci-agent/internal/executor/metrics	0.075s
?   	github.com/cirruslabs/cirrus-ci-agent/internal/executor/metrics/source	[no test files]
?   	github.com/cirruslabs/cirrus-ci-agent/internal/executor/metrics/source/cgroup	[no test files]
?   	github.com/cirruslabs/cirrus-ci-agent/internal/executor/metrics/source/cgroup/cpu	[no test files]
?   	github.com/cirruslabs/cirrus-ci-agent/internal/executor/metrics/source/cgroup/memory	[no test files]
ok  	github.com/cirruslabs/cirrus-ci-agent/internal/executor/metrics/source/cgroup/parser	(cached)
?   	github.com/cirruslabs/cirrus-ci-agent/internal/executor/metrics/source/cgroup/resolver	[no test files]
?   	github.com/cirruslabs/cirrus-ci-agent/internal/executor/metrics/source/cgroup/subsystem	[no test files]
?   	github.com/cirruslabs/cirrus-ci-agent/internal/executor/metrics/source/system	[no test files]
?   	github.com/cirruslabs/cirrus-ci-agent/internal/executor/piper	[no test files]
?   	github.com/cirruslabs/cirrus-ci-agent/internal/executor/processdumper	[no test files]
ok  	github.com/cirruslabs/cirrus-ci-agent/internal/executor/terminalwrapper	(cached)
?   	github.com/cirruslabs/cirrus-ci-agent/internal/executor/updatebatcher	[no test files]
ok  	github.com/cirruslabs/cirrus-ci-agent/internal/hasher	(cached)
?   	github.com/cirruslabs/cirrus-ci-agent/internal/http_cache	[no test files]
ok  	github.com/cirruslabs/cirrus-ci-agent/internal/network	(cached)
?   	github.com/cirruslabs/cirrus-ci-agent/internal/shellwords	[no test files]
?   	github.com/cirruslabs/cirrus-ci-agent/internal/signalfilter	[no test files]
ok  	github.com/cirruslabs/cirrus-ci-agent/internal/targz	(cached)
?   	github.com/cirruslabs/cirrus-ci-agent/internal/testutil	[no test files]
ok  	github.com/cirruslabs/cirrus-ci-agent/pkg/grpchelper	(cached)
FAIL
```
On **NetBSD**:
```
# github.com/mitchellh/go-ps
../go/pkg/mod/github.com/mitchellh/go-ps@v1.0.0/process.go:31:9: undefined: processes
../go/pkg/mod/github.com/mitchellh/go-ps@v1.0.0/process.go:39:9: undefined: findProcess
?       github.com/cirruslabs/cirrus-ci-agent/api       [no test files]
ok      github.com/cirruslabs/cirrus-ci-agent/cmd/agent (cached)
?       github.com/cirruslabs/cirrus-ci-agent/cmd/metrics       [no test files]
ok      github.com/cirruslabs/cirrus-ci-agent/internal/cirrusenv        (cached)
?       github.com/cirruslabs/cirrus-ci-agent/internal/client   [no test files]
FAIL    github.com/cirruslabs/cirrus-ci-agent/internal/executor [build failed]
Error #0: failed to query CPU usage using gopsutil on netbsd/amd64: not implemented yet
Error #1: failed to query memory usage using gopsutil on netbsd/amd64: not implemented yet
--- FAIL: TestMetrics (4.52s)
    metrics_test.go:23: 
                Error Trace:    metrics_test.go:23
                Error:          Should be empty, but was [failed to query CPU usage using gopsutil on netbsd/amd64: not implemented yet failed to query memory usage using gopsutil on netbsd/amd64: not implemented yet]
                Test:           TestMetrics
                Messages:       we should never get errors here, but got 2
FAIL
FAIL    github.com/cirruslabs/cirrus-ci-agent/internal/executor/metrics 4.529s
?       github.com/cirruslabs/cirrus-ci-agent/internal/executor/metrics/source  [no test files]
?       github.com/cirruslabs/cirrus-ci-agent/internal/executor/metrics/source/cgroup   [no test files]
?       github.com/cirruslabs/cirrus-ci-agent/internal/executor/metrics/source/cgroup/cpu       [no test files]
?       github.com/cirruslabs/cirrus-ci-agent/internal/executor/metrics/source/cgroup/memory    [no test files]
ok      github.com/cirruslabs/cirrus-ci-agent/internal/executor/metrics/source/cgroup/parser    (cached)
?       github.com/cirruslabs/cirrus-ci-agent/internal/executor/metrics/source/cgroup/resolver  [no test files]
?       github.com/cirruslabs/cirrus-ci-agent/internal/executor/metrics/source/cgroup/subsystem [no test files]
?       github.com/cirruslabs/cirrus-ci-agent/internal/executor/metrics/source/system   [no test files]
?       github.com/cirruslabs/cirrus-ci-agent/internal/executor/piper   [no test files]
?       github.com/cirruslabs/cirrus-ci-agent/internal/executor/processdumper   [no test files]
ok      github.com/cirruslabs/cirrus-ci-agent/internal/executor/terminalwrapper (cached)
?       github.com/cirruslabs/cirrus-ci-agent/internal/executor/updatebatcher   [no test files]
ok      github.com/cirruslabs/cirrus-ci-agent/internal/hasher   (cached)
?       github.com/cirruslabs/cirrus-ci-agent/internal/http_cache       [no test files]
ok      github.com/cirruslabs/cirrus-ci-agent/internal/network  (cached)
?       github.com/cirruslabs/cirrus-ci-agent/internal/shellwords       [no test files]
?       github.com/cirruslabs/cirrus-ci-agent/internal/signalfilter     [no test files]
ok      github.com/cirruslabs/cirrus-ci-agent/internal/targz    (cached)
?       github.com/cirruslabs/cirrus-ci-agent/internal/testutil [no test files]
ok      github.com/cirruslabs/cirrus-ci-agent/pkg/grpchelper    (cached)
FAIL

```


Could you add a support for  OpenBSD / NetBSD?

Thanks!

--- a/internal/executor/metrics/metrics.go
+++ b/internal/executor/metrics/metrics.go
@@ -1,4 +1,4 @@
-//go:build !openbsd
+//go:build !(openbsd || netbsd)
 
 package metrics
 

--- a/internal/executor/metrics/metrics_test.go
+++ b/internal/executor/metrics/metrics_test.go
@@ -1,3 +1,5 @@
+//go:build !(openbsd || netbsd)
+
 package metrics_test
 
 import (

--- a/internal/executor/metrics/metrics_unsupported.go
+++ b/internal/executor/metrics/metrics_unsupported.go
@@ -1,4 +1,4 @@
-//go:build openbsd
+//go:build openbsd || netbsd
 
 package metrics
 

--- a/internal/executor/shell_subunix_test.go
+++ b/internal/executor/shell_subunix_test.go
@@ -1,0 +1,46 @@
+//go:build unix && !(openbsd || netbsd)
+
+package executor
+
+import (
+	"context"
+	"github.com/mitchellh/go-ps"
+	"github.com/stretchr/testify/assert"
+	"regexp"
+	"strconv"
+	"testing"
+	"time"
+)
+
+// TestProcessGroupTermination ensures that we terminate the whole process group that
+// the shell spawned in ShellCommandsAndGetOutput() has been placed into, thus killing
+// it's children processes.
+func TestProcessGroupTermination(t *testing.T) {
+	ctx, cancel := context.WithTimeout(context.Background(), 1*time.Second)
+	defer cancel()
+
+	success, output := ShellCommandsAndGetOutput(ctx, []string{"sleep 86400 & echo target PID is $! ; sleep 60"}, nil)
+
+	assert.False(t, success, "the command should fail due to time out error")
+	assert.Contains(t, output, "Timed out!", "the command should time out")
+
+	re := regexp.MustCompile(".*target PID is ([0-9]+).*")
+	matches := re.FindStringSubmatch(output)
+	if len(matches) != 2 {
+		t.Fatal("failed to find target PID")
+	}
+
+	pid, err := strconv.ParseInt(matches[1], 10, 32)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// Wait for the zombie to be reaped by the init process
+	time.Sleep(5 * time.Second)
+
+	// Unfortunately go-ps error behavior differs depending on the OS,
+	// (missing process is an error on FreeBSD but there's no error on Linux),
+	// so we skip the check here
+	process, _ := ps.FindProcess(int(pid))
+	assert.Nil(t, process)
+}

--- a/internal/executor/shell_unix_test.go
+++ b/internal/executor/shell_unix_test.go
@@ -5,47 +5,11 @@ package executor
 
 import (
 	"context"
-	"github.com/mitchellh/go-ps"
 	"github.com/stretchr/testify/assert"
-	"regexp"
 	"runtime"
-	"strconv"
 	"testing"
 	"time"
 )
-
-// TestProcessGroupTermination ensures that we terminate the whole process group that
-// the shell spawned in ShellCommandsAndGetOutput() has been placed into, thus killing
-// it's children processes.
-func TestProcessGroupTermination(t *testing.T) {
-	ctx, cancel := context.WithTimeout(context.Background(), 1*time.Second)
-	defer cancel()
-
-	success, output := ShellCommandsAndGetOutput(ctx, []string{"sleep 86400 & echo target PID is $! ; sleep 60"}, nil)
-
-	assert.False(t, success, "the command should fail due to time out error")
-	assert.Contains(t, output, "Timed out!", "the command should time out")
-
-	re := regexp.MustCompile(".*target PID is ([0-9]+).*")
-	matches := re.FindStringSubmatch(output)
-	if len(matches) != 2 {
-		t.Fatal("failed to find target PID")
-	}
-
-	pid, err := strconv.ParseInt(matches[1], 10, 32)
-	if err != nil {
-		t.Fatal(err)
-	}
-
-	// Wait for the zombie to be reaped by the init process
-	time.Sleep(5 * time.Second)
-
-	// Unfortunately go-ps error behavior differs depending on the OS,
-	// (missing process is an error on FreeBSD but there's no error on Linux),
-	// so we skip the check here
-	process, _ := ps.FindProcess(int(pid))
-	assert.Nil(t, process)
-}
 
 func TestZshDoesNotHang(t *testing.T) {
 	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)

--- a/internal/executor/shell_unix_test.go
+++ b/internal/executor/shell_unix_test.go
@@ -6,6 +6,7 @@ package executor
 import (
 	"context"
 	"github.com/stretchr/testify/assert"
+	"os/exec"
 	"runtime"
 	"testing"
 	"time"
@@ -14,6 +15,10 @@ import (
 func TestZshDoesNotHang(t *testing.T) {
 	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
 	defer cancel()
+
+	if _, err := exec.LookPath("zsh"); err != nil {
+		t.Skip("no Zsh found")
+	}
 
 	success, _ := ShellCommandsAndGetOutput(ctx, []string{"zsh -c 'echo \"a:b\" | read -d \":\" piece'"}, nil)
 	assert.True(t, success)

